subsys/bluetooth/host_extensions/custom_ltk.c

```diff
@@ -0,0 +1,109 @@
+#include <bluetooth/nrf/host_extensions.h>
+#include <errno.h>
+#include <stdint.h>
+#include <string.h>
+#include <zephyr/bluetooth/bluetooth.h>
+#include <zephyr/bluetooth/conn.h>
+#include <zephyr/sys/check.h>
+#include <zephyr/sys/util_macro.h>
+#include <zephyr/toolchain.h>
+
+#include "../subsys/bluetooth/host/keys.h"
+#include "../subsys/bluetooth/host/conn_internal.h"
+#include "zephyr/kernel.h"
+
+enum {
+	NRF_BT_KEYS_TRANSIENT = BIT(7),
+};
+
+int nrf_bt_conn_set_ltk(struct bt_conn *conn, const struct nrf_ltk *ltk, bool authenticated)
+{
+	struct bt_keys *keys;
+
+	CHECKIF(conn == NULL || ltk == NULL) {
+		return -EINVAL;
+	}
+
+	/* This hopefully a crude but decent prevention of any race
+	 * conditions with smp.c.
+	 *
+	 * It should prevent a race with smp_security_request(),
+	 * smp_security_request() and bt_smp_start_security(), which
+	 * will all allocate keys early and cause the following code
+	 * to find a key and return -EALREADY.
+	 */
+	k_sched_lock();
+
+	/* If there is a prior pairing for this address, we don't
+	 * allow setting the LTK manually. Please unpair first.
+	 *
+	 *
+	 */
+	keys = bt_keys_find(BT_KEYS_ALL, conn->id, &conn->le.dst);
+	if (keys) {
+		return -EALREADY;
+	}
+
+	/* We are using "get" variant here so it allocates.
+	 */
+	keys = bt_keys_get_addr(conn->id, &conn->le.dst);
+	if (!keys) {
+		return -ENOMEM;
+	}
+
+	conn->le.keys = keys;
+
+	/* Use the LESC LTK type, which is symmetric for central and
+	 * peripheral.
+	 */
+	keys->keys = BT_KEYS_LTK_P256;
+
+	memcpy(keys->ltk.val, ltk->val, sizeof(ltk->val));
+	keys->enc_size = sizeof(ltk->val);
+
+	keys->flags = 0;
+	keys->flags |= BT_KEYS_SC;
+	keys->flags |= NRF_BT_KEYS_TRANSIENT;
+
+	if (authenticated) {
+		keys->flags |= BT_KEYS_AUTHENTICATED;
+	}
+
+	k_sched_unlock();
+
+	return 0;
+}
+
+static void _clear_keys_if_transient(struct bt_conn *conn)
+{
+	struct bt_keys *keys = conn->le.keys;
+
+	if (keys && keys->flags & NRF_BT_KEYS_TRANSIENT) {
+		/* Clear the LTK so that this bt_keys looks like it
+		 * belongs to a paired not bonded device, and the
+		 * bt_keys is not saved to flash. Clearing the key is
+		 * also good for security.
+		 */
+		keys->ltk = (struct bt_ltk){};
+		keys->keys = 0;
+	}
+}
+
+static void _on_disconnect(struct bt_conn *conn, uint8_t reason)
+{
+	ARG_UNUSED(reason);
+	_clear_keys_if_transient(conn);
+}
+
+static void _on_security_changed(struct bt_conn *conn, bt_security_t level,
+				 enum bt_security_err err)
+{
+	ARG_UNUSED(level);
+	ARG_UNUSED(err);
+	_clear_keys_if_transient(conn);
```

Comment:
Do we want to clear this on security changed? The bond will no longer show up after this, i. e. when using the bt_foreach_bond function.